### DISCUSSION TITLE
Fun with  `shapely` instersections...

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#500] [Show and tell]
![sketch_2024_08_30](https://github.com/user-attachments/assets/32f190ba-141d-4942-a4e2-6fdbc8c30db9)

```python
from itertools import combinations
from functools import lru_cache

import py5
from shapely.geometry import Polygon, GeometryCollection, Point
from shapely.affinity import translate as shapely_translate

polys = [
    Point((200, 200)).buffer(100),
    Point((300, 200)).buffer(100),
    Point((300, 300)).buffer(100),
    Point((200, 300)).buffer(100),
]

dragged_poly = -1  # no dragged poly

def setup():
    py5.size(500, 500)
    py5.color_mode(py5.HSB)

def predraw_update():
    global shp
    rs = get_regions(tuple(polys))
    shp = py5.convert_cached_shape(rs)
    # TODO maybe replace this with a recursive function
    flat_children = []
    if shp.get_child_count():  
        children = shp.get_children()
        for child in children:
            if child.get_child_count():  # there are grandchildren
                flat_children.extend(child.get_children())
            else:
                flat_children.append(child)
    # Change colors
    for i, child in enumerate(flat_children):
        # Py5Shape .contains() for mouse over darkening!
        b = 128 if child.contains(py5.mouse_x, py5.mouse_y) else 255
        child.set_fill(py5.color((i * 18) % 255, 255, b))

def draw():
    py5.background(32, 50, 200)
    py5.shape(shp, 0, 0)

def mouse_pressed():
    global dragged_poly
    for i, p in enumerate(polys):
        if p.contains(Point((py5.mouse_x, py5.mouse_y))):
            dragged_poly = i
            break
            
def mouse_released():
    global dragged_poly
    dragged_poly = -1  # no dragged poly    

def mouse_dragged():
    if dragged_poly >= 0:
        dx = py5.mouse_x - py5.pmouse_x
        dy = py5.mouse_y - py5.pmouse_y
        polys[dragged_poly] = shapely_translate(polys[dragged_poly], dx, dy)
    
@lru_cache(maxsize=64)
def get_regions(polys):
    global regions # for debugging
    regions = []
    n = len(polys)
    for i in range(1, n + 1):
        for combo in combinations(range(n), i):
            intersection = polys[combo[0]]
            for j in combo[1:]:    # get the intersections
                intersection = intersection.intersection(polys[j])
            # Subtract non-combo parts (I like without this, it's fun too!)
            intersection = subtract_others(intersection, combo, polys)
            # Flatten MultiPolygons... this was not enough somehow to fix the coloring.
            if intersection.area:
                if intersection.geom_type == 'Polygon':
                    regions.append(intersection)
                else:  # I was using regions.extend(intersection.geoms) but decided to check areas again 
                    for element in intersection.geoms:
                        if element.area:
                            regions.append(element)
        
    return GeometryCollection(tuple(r.buffer(-3) for r in regions))

def subtract_others(intersection, combo, polys):
    for k in range(len(polys)):
        if k not in combo:
            intersection = intersection.difference(polys[k])
    return intersection

py5.run_sketch(block=False)
```

**Comment by hx2A** [comment]
This is great! Thanks for sharing this. I like the use of `predraw_update()` here. How much of a difference does it make for performance? Shapely works very well here for what you are doing.

**Reply by villares** [reply]
I'm happy you enjoyed it! As for the speed gains
... I'd  have to profile it...

**Reply by hx2A** [reply]
For the speed gains, I'd just check the frame rate with and without this feature.

**Reply by villares** [reply]
adding more circles, and dragging one of them madly for several seconds I got frame rates close to 30 either with `predraw_update()` or without it, so the gain was not very noticeable... Releasing it gets back to around 58.

![image](https://github.com/user-attachments/assets/d298881c-3cdc-4417-84db-91c926cdfdb5)

**Reply by villares** [reply]
With yet more circles it can get as slow as 12 on drag, but I again I couldn't notice much difference... maybe a tad faster with predraw.
![image](https://github.com/user-attachments/assets/20b1436f-649d-4111-bbbb-b7b73d9aa9c4)

**Reply by hx2A** [reply]
> maybe a tad faster with predraw.

That's about what I expected. To measure the difference scientifically you'd need to run both versions at the same time and measure statistics of the frame rate over time, and then maybe you'd see a few % difference in the frame rate. It isn't much, and not the performance gain I was hoping for when I wrote the code for that feature.

**Reply by villares** [reply]
Things I want to try at some point... 
- See if I can think of ways of improving this, maybe with `shapely.symetric_difference`
- How would you add custom data to shapely objects, like, maybe color or "parents"?

**Reply by hx2A** [reply]
> Things I want to try at some point...
> 
>     * See if I can think of ways of improving this, maybe with `shapely.symetric_difference`

How about each circle has its own color, and the color drawn is a mixture of the colors?

You can improve the code by using Shapely's [set operations](https://shapely.readthedocs.io/en/stable/set_operations.html) such as `intersection_all()`. The performance would be faster.

>     * How would you add custom data to shapely objects, like, maybe color or "parents"?

I recall the shapely documentation saying you can use assignments to add custom data to Shapely objects, such as `s.custom_property = 42`. That wouldn't progate through boolean operations though.

**Reply by villares** [reply]
Oh, the set operations are nice, I have used also `unary_union` before.
Yes, the custom data would be to add colors, and I figured I'd have to plug them in after each operation...

**Reply by villares** [reply]
Custom attributes are no longer permitted...
![image](https://github.com/user-attachments/assets/b52440ea-2ac5-4554-94d4-c0153015115a)
https://shapely.readthedocs.io/en/latest/migration.html#setting-custom-attributes

**Reply by hx2A** [reply]
Ha! I knew I read about that feature somewhere, but forgot it went away. Bummer! Well, since the objects are immutable, you can use them as keys in a dictionary and store whatever values you want there.

**Reply by villares** [reply]
yeah, and the immutable thing is what allowed us to have the caching (memoization with dictionaries usually).

**Comment by villares** [comment]
Probably faster, but but also more readable!

```python
@lru_cache(maxsize=64)
def get_regions(polys):
    global regions # for debugging
    regions = []
    n = len(polys)
    for i in range(1, n + 1):
        for combo in combinations(range(n), i):
            intersection = intersection_all([polys[j] for j in combo])
            not_in_combo = unary_union([polys[k] for k in range(n) if k not in combo])
            regions.append(intersection.difference(not_in_combo))
    return GeometryCollection(regions).buffer(-3)
```
The only thing is now the resulting polygons order (and thus coloring on the example) is less stable, changing at each small move.

**Reply by villares** [reply]
![sketch_2024_09_01](https://github.com/user-attachments/assets/2bbdd04d-fcc3-462d-a5df-eff23db58b28)

I like this!

```python
from itertools import combinations
from functools import lru_cache

import py5
from shapely.geometry import Polygon, GeometryCollection, Point
from shapely.affinity import translate as shapely_translate
from shapely import intersection_all, unary_union

polys = [
    Point((200, 200)).buffer(100),
    Point((300, 200)).buffer(100),
    Point((300, 300)).buffer(100),
    Point((200, 300)).buffer(100),
]

dragged_poly = -1  # no dragged poly

def setup():
    py5.size(500, 500)
    py5.color_mode(py5.HSB)

def predraw_update():
    global group_shp  #, shp, children, rs, poly
    rs = get_regions(tuple(polys))
    group_shp = py5.create_shape(py5.GROUP)
    for poly in rs.geoms:
        shp = py5.convert_cached_shape(poly)
        b = 128 if shp.contains(py5.mouse_x, py5.mouse_y) else 255
        shp.set_fill(py5.color(int(poly.area / 100) % 255, 255, b))
        group_shp.add_child(shp)
    
def draw():
    py5.background(32, 50, 200)
    py5.shape(group_shp, 0, 0)
    py5.window_title(f'{py5.get_frame_rate():.1f}')

def mouse_pressed():
    global dragged_poly
    for i, p in enumerate(polys):
        if p.contains(Point((py5.mouse_x, py5.mouse_y))):
            dragged_poly = i
            break
            
def mouse_released():
    global dragged_poly
    dragged_poly = -1  # no dragged poly    

def mouse_dragged():
    if dragged_poly >= 0:
        dx = py5.mouse_x - py5.pmouse_x
        dy = py5.mouse_y - py5.pmouse_y
        polys[dragged_poly] = shapely_translate(polys[dragged_poly], dx, dy)
    
@lru_cache(maxsize=64)
def get_regions(polys):
    global regions # for debugging
    regions = []
    n = len(polys)
    for i in range(1, n + 1):
        for combo in combinations(range(n), i):
            intersection = intersection_all([polys[j] for j in combo])
            not_in_combo = unary_union([polys[k] for k in range(n) if k not in combo])
            regions.append(intersection.difference(not_in_combo))
    return GeometryCollection(regions).buffer(-3)

py5.run_sketch(block=False)
```

**Reply by hx2A** [reply]
Neat and organized code!

If you wanted the colors to be more stable, you'd need to assign colors to each of the Polygons in `polys` and then have a deterministic algorithm for picking a new color based on the colors of the Polygons involved in the intersection.

**Reply by villares** [reply]
It was lucky I could simplify it removing most checks and filters because the better results of the bulk methods.

Exactly, the next thing I'll will try is to assign the color based on the combination of parents as a custom proprety... 

Crazy integration idea... maybe good as a third-party add-on, if one registers *geodataframes* for the `convert_shape()` and we could check for a *py5_fill_color* column and other attributes...

**Reply by hx2A** [reply]
> Crazy integration idea... maybe good as a third-party add-on, if one registers _geodataframes_ for the `convert_shape()` and we could check for a _py5_fill_color_ column and other attributes...

This is a great idea! The ability to register new conversion functionality is part of the design of py5's code integration features. This would be super useful. Perhaps it could create a group shape with child shapes for each row, and look for certain column names such as _py5_fill_color_ or _py5_stroke_weight_, and if not set, just use py5's current style information? I don't think it would be that hard to make. I think it will just be a for loop to iterate through the rows, some if statements, and then calls to the existing shapely integration code.
